crates/starknet-server/src/api/json_rpc/write_endpoints.rs

```diff
@@ -0,0 +1,153 @@
+            ))
+        })?;
+
+        Ok(ContractClass::Cairo0(starknet_types::contract_class::Cairo0ContractClass::Json(
+            json!({
+                "program": value.program,
+                "abi": abi_json,
+                "entry_points_by_type": entry_points_json,
+            }),
+        )))
+    }
+}
+
+impl TryFrom<BroadcastedDeclareTransactionV1>
+    for starknet_core::transactions::declare_transaction::DeclareTransactionV1
+{
+    type Error = ApiError;
+    fn try_from(value: BroadcastedDeclareTransactionV1) -> RpcResult<Self> {
+        Ok(DeclareTransactionV1 {
+            sender_address: value.sender_address.0,
+            max_fee: value.common.max_fee.0,
+            signature: value.common.signature.iter().map(|x| x.0).collect(),
+            nonce: value.common.nonce.0,
+            contract_class: ContractClass::try_from(value.contract_class)?,
+            class_hash: None,
+            transaction_hash: None,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use starknet_core::{Starknet, StarknetConfig};
+    use starknet_types::traits::ToHexString;
+
+    use crate::api::json_rpc::JsonRpcHandler;
+    use crate::api::models::transaction::BroadcastedDeclareTransactionV1;
+    use crate::api::Api;
+
+    #[tokio::test]
+    async fn add_declare_transaction_v1_should_be_successful() {
+        let declare_txn_v1 = test_broadcasted_declare_transaction_v1();
+        let json_rpc_handler = setup();
+        let result = json_rpc_handler
+            .add_declare_transaction(
+                crate::api::models::transaction::BroadcastedDeclareTransaction::V1(Box::new(
+                    declare_txn_v1.clone(),
+                )),
+            )
+            .await
+            .unwrap();
+
+        // Data taken from transaction execution to https://alpha4.starknet.io/gateway/add_transaction
+        // which resulted in transaction_hash
+        // 0x1d50d192f54d8d75e73c8ab8fb7159e70bfdbccc322abb43a081889a3043627 Could be checked in https://testnet.starkscan.co/tx/0x1d50d192f54d8d75e73c8ab8fb7159e70bfdbccc322abb43a081889a3043627
+        assert_eq!(
+            result.class_hash.0.to_prefixed_hex_str(),
+            "0x399998c787e0a063c3ac1d2abac084dcbe09954e3b156d53a8c43a02aa27d35"
+        );
+
+        println!("{}", result.transaction_hash.0.to_prefixed_hex_str());
+    }
+
+    fn setup() -> JsonRpcHandler {
+        let config = StarknetConfig {
+            seed: 123,
+            total_accounts: 1,
+            predeployed_accounts_initial_balance: 1000.into(),
+        };
+        let starknet = Starknet::new(&config).unwrap();
+        let api = Api::new(starknet);
+        JsonRpcHandler { api }
+    }
+
+    #[test]
+    fn parsed_base64_gzipped_json_contract_class_correctly_computed_hash() {
+        let json_string = std::fs::read_to_string(concat!(
+            env!("CARGO_MANIFEST_DIR"),
+            "/test_data/rpc/declare_v1.json"
```

Comment:
And what are we declaring? Should it be documented somewhere? Right now it's a hardcoded artifact and it is not clear what is the source.

Reply:
Left a comment in [7c72a10](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/39/commits/7c72a107d1376d8af48e814da62c52f4af76fd26)